ibis/backends/dask/udf.py

```diff
@@ -63,29 +63,29 @@ def execute_udf_node_groupby(op, *args, **kwargs):
     @execute_node.register(
         ops.ElementWiseVectorizedUDF, *(itertools.repeat(dd.Series, nargs))
     )
-    def execute_udf_node(op, *args, **kwargs):
+    def execute_udf_node(op, *args, cache=None, **kwargs):
         # We have rewritten op.func to be a closure enclosing
         # the kwargs, and therefore, we do not need to pass
         # kwargs here. This is true for all udf execution in this
         # file.
         # See ibis.udf.vectorized.UserDefinedFunction
+        try:
+            return cache[op]
```

Comment:
Can you add some comments here explaining how the cache prevent struct UDF re-execution - Does the struct UDF first gets executed in here then the StructField ops just read from the cache?

Reply:
Also, does it make sense to cache UDF that is not Struct output? Can those be reused as well? 

Thinking about memory foot print too - in a UDF heavy expression, every UDF execution get cached, which means the total memory footprint could be quite higher than before? (effectively all UDF results are stored twice in memory, one in the cache and one in the DataFrame)?

Reply:
It's probably true that the pandas backend will use extra memory given the copying that it does. There's not much we can do there without a rewrite of that backend which isn't in scope at the moment (unless someone wants to volunteer for that!)

Hard to say what happens in the dask case without measuring a representative workload.

I think we should address this later, if it even shows up.

I think @kszucs has done a great job of addressing the original concern of avoiding re-execution of UDFs.

Reply:
That should be the current behaviour already since we don't check for struct datatypes anywhere.

Reply:
>It's probably true that the pandas backend will use extra memory given the copying that it does. There's not much we can do there without a rewrite of that backend which isn't in scope at the moment (unless someone wants to volunteer for that!)

Yeah I won't object to move forward with this. But should probably document these limitations.

Reply:
> Possibly. That's a good follow-up issue.

What I meant is for non Struct output these cache probably won't be reused so it's kind of like a waste of memory but I haven't really thought this through.

Reply:
We should extend the UDF operations to mark them idempotent, [see how spark handles deterministic VS non-deterministic UDFs](https://spark.apache.org/docs/3.1.3/api/python/reference/api/pyspark.sql.functions.udf.html). DataFusion uses the [volatility enum](https://github.com/apache/arrow-datafusion/blob/41b4e491663029f653e491b110d0b5e74d08a0b6/datafusion/expr/src/signature.rs#L25-L37) for this purpose.

Reply:
Created a follow-up issue to track this https://github.com/ibis-project/ibis/issues/4355

Reply:
Sounds good to me